### PR TITLE
Use original branch token instead of deserializing as branch info and then re-serializing

### PR DESCRIPTION
**Pull request by norberthu**
This fixes the branch token persisted in history tree info as being
potentially lossy. This change ensures forward/backward compatibility
by writing both branch info as well as the branch token. On read, the
logic fallbacks to branch info if branch token is not found.

The double write to the deprecated branch info may be removed in the
future, afterwhich no running temporal service with data may be
downgraded to "before" this change with fallback reads.

Similarly, the fallback read may be removed in the future, afterwhich
no running temporal service with data may be upgraded from "before"
this change with double writes.

**What changed?**

**Why?**

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: common/persistence/history_manager.go [hunk 1]
```diff
@@ -361,6 +372,35 @@ func (m *executionManagerImpl) getHistoryBranchInfo(
...
+	// compute the missing branch token on the fly. Similarly, if only opaque branch token exists but without the branch
+	// info, then compute the missing branch info on the fly.
+	// TODO: Consider removing in the future, but this will be a forward/backward incompatible change for old data.
+	if treeInfo.BranchToken == nil {
+		blob, err := serializer.HistoryBranchToBlob(treeInfo.BranchInfo, enumspb.ENCODING_TYPE_PROTO3)
+		if err != nil {
+			return nil, err
+		}
+		treeInfo.BranchToken = blob.Data
+	} else if treeInfo.BranchInfo == nil {
```
Comment (yycptt) [inline common/persistence/history_manager.go:390]: Why we need fill in the `BranchInfo`? I thought all new logic is going to use BranchToken field? and we do double write to be compatible with the old code.

File: service/worker/scanner/history/scavenger.go [hunk 2]
```diff
@@ -333,15 +324,13 @@ func getTaskLoggingTags(err error, task taskDetail) []tag.Tag {
 			tag.WorkflowNamespaceID(task.namespaceID),
 			tag.WorkflowID(task.workflowID),
 			tag.WorkflowRunID(task.runID),
-			tag.WorkflowTreeID(task.treeID),
-			tag.WorkflowBranchID(task.branchID),
+			tag.WorkflowBranchToken(task.branchToken),
```
Comment (yycptt) [inline service/worker/scanner/history/scavenger.go:327]: I am not sure how useful this is without having it decoded. we also can't filter on this field basically...

Comment (norberthu) [reply service/worker/scanner/history/scavenger.go:327]: I merely followed the same pattern as `tag.ArchivalRequestBranchToken()`. If you're okay with dropping this, I can just remove it.

File: common/persistence/history_manager.go (hunk 1 above)
Comment (norberthu) [reply common/persistence/history_manager.go:390]: I tried to describe this in the git message body. Basically, if some customer were to upgrade a live deployment from before this change to after this change, then they wouldn't be able to downgrade unless we double write. At some point (in the medium~distant future) this double write will be removed and customers won't be able to downgrade to before this change.

**Review by yycptt — APPROVED** [review]

File: service/worker/scanner/history/scavenger.go (hunk 2 above)
Comment (yycptt) [reply service/worker/scanner/history/scavenger.go:327]: Let's keep it. I was originally thinking maybe in error case we should decode the branch token and use the original treeID and branchID tag so during debugging we can filter on these tags. 
But second thought on this, we also only have branchToken in mutable state, not tree/branchID. So filtering on branchToken will actually make things easier.

**Pull request merged**